? debian/diff/parisc-sendto-recvfrom.diff
Index: parisc/recvfrom.S
===================================================================
RCS file: /cvs/dietlibc/parisc/recvfrom.S,v
retrieving revision 1.1
diff -u -r1.1 recvfrom.S
--- parisc/recvfrom.S	20 Jul 2002 13:29:27 -0000	1.1
+++ parisc/recvfrom.S	20 Dec 2005 12:29:14 -0000
@@ -1,3 +1,12 @@
 #include "syscalls.h"
 
-syscall(recvfrom, recvfrom);
+.level 1.1
+.globl recvfrom
+.type recvfrom,@function
+recvfrom:
+	/* Just load the arguments from the
+	   stack into %r21 and %r22  */
+	ldw -56(%sr0, %sp), %r21 ; 0x78
+	ldw -52(%sr0, %sp), %r22 ; 0x74
+	b __unified_syscall
+	ldi __NR_recvfrom, %r20
Index: parisc/sendto.S
===================================================================
RCS file: /cvs/dietlibc/parisc/sendto.S,v
retrieving revision 1.1
diff -u -r1.1 sendto.S
--- parisc/sendto.S	20 Jul 2002 13:29:27 -0000	1.1
+++ parisc/sendto.S	20 Dec 2005 12:29:15 -0000
@@ -1,3 +1,12 @@
 #include "syscalls.h"
 
-syscall(sendto, sendto);
+.level 1.1
+.globl sendto
+.type sendto,@function
+sendto:
+	/* Just load the arguments from the
+	   stack into %r21 and %r22  */
+	ldw -56(%sr0, %sp), %r21 ; 0x78
+	ldw -52(%sr0, %sp), %r22 ; 0x74
+	b __unified_syscall
+	ldi __NR_sendto, %r20
