Index: librpc/svc_auth_unix.c
===================================================================
RCS file: /var/lib/cvs/dietlibc/librpc/svc_auth_unix.c,v
retrieving revision 1.2
diff -u -r1.2 svc_auth_unix.c
--- librpc/svc_auth_unix.c	13 Sep 2003 08:46:46 -0000	1.2
+++ librpc/svc_auth_unix.c	7 Dec 2004 16:14:26 -0000
@@ -51,7 +51,7 @@
 /*
  * Unix longhand authenticator
  */
-static enum auth_stat _svcauth_unix(rqst, msg)
+enum auth_stat _svcauth_unix(rqst, msg)
 register struct svc_req *rqst;
 register struct rpc_msg *msg;
 {
@@ -128,7 +128,7 @@
  * Shorthand unix authenticator
  * Looks up longhand in a cache.
  */
- /*ARGSUSED*/ static enum auth_stat _svcauth_short(rqst, msg)
+ /*ARGSUSED*/ enum auth_stat _svcauth_short(rqst, msg)
 struct svc_req *rqst;
 struct rpc_msg *msg;
 {
