Index: Makefile
===================================================================
RCS file: /cvs/dietlibc/Makefile,v
retrieving revision 1.182
diff -u -r1.182 Makefile
--- Makefile	7 May 2006 04:16:59 -0000	1.182
+++ Makefile	13 Jun 2006 13:12:32 -0000
@@ -262,7 +262,7 @@
 DYN_LIBC_PIC = $(LIBOBJ) $(LIBSTDIOOBJ) $(LIBUGLYOBJ) \
 $(LIBCRUFTOBJ) $(LIBCRYPTOBJ) $(LIBSHELLOBJ) $(LIBREGEXOBJ)
 
-DYN_LIBC_OBJ = $(PICODIR)/dyn_syscalls.o $(PICODIR)/errlist.o \
+DYN_LIBC_OBJ = $(PICODIR)/dyn_syscalls.o $(PICODIR)/errlist.o $(PICODIR)/__guard.o \
 	$(patsubst $(OBJDIR)/%.o,$(PICODIR)/%.o,$(DYN_LIBC_PIC))
 
 DYN_PTHREAD_OBJS = $(patsubst $(OBJDIR)/%.o,$(PICODIR)/%.o,$(LIBPTHREAD_OBJS))
