From Ivan Jager (#604717)

Index: dietlibc-0.33~cvs20110918/include/endian.h
===================================================================
--- dietlibc-0.33~cvs20110918.orig/include/endian.h	2011-09-16 20:34:24.000000000 +0200
+++ dietlibc-0.33~cvs20110918/include/endian.h	2011-11-03 21:52:13.000000000 +0100
@@ -39,7 +39,7 @@
 # define __LONG_LONG_PAIR(HI, LO) HI, LO
 #endif
 
-#if defined(__alpha__) || defined(__mips64__) || defined(__sparc_v9__) || defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__s390x__)
+#if defined(__alpha__) || defined(__mips64__) || defined(__sparcv9) || defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__s390x__) || defined(__arch64__)
 #define __WORDSIZE 64
 #endif
 
