From Ivan Jager (#604717)

Index: pkg-dietlibc/include/endian.h
===================================================================
--- pkg-dietlibc.orig/include/endian.h	2012-03-25 15:07:33.000000000 +0000
+++ pkg-dietlibc/include/endian.h	2012-03-25 15:12:47.000000000 +0000
@@ -39,7 +39,7 @@
 # define __LONG_LONG_PAIR(HI, LO) HI, LO
 #endif
 
-#if defined(__alpha__) || defined(__mips64__) || defined(__sparc_v9__) || defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__s390x__)
+#if defined(__alpha__) || defined(__mips64__) || defined(__sparcv9) || defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__s390x__) || defined(__arch64__)
 #define __WORDSIZE 64
 #endif
 
