--- dietlibc-0.32/diet.c	2011-02-24 23:59:04.000000000 +0000
+++ dietlibc-0.32.fix/diet.c	2011-02-24 23:58:28.000000000 +0000
@@ -33,10 +33,14 @@
   "sparc64","-Os","-m64","-mhard-quad-float",0,
   "alpha","-Os","-fomit-frame-pointer",0,
 #ifdef __ARM_EABI__
-  "arm","-Os","-fomit-frame-pointer","-mfloat-abi=soft","-meabi=4",0,
+#ifdef __ARM_PCS_VFP__
+  "arm","-Os","-fomit-frame-pointer","-mfloat-abi=hard",0,
+#else
+  "arm","-Os","-fomit-frame-pointer","-mfloat-abi=soft",0,
+#endif
 #else
   "arm","-Os","-fomit-frame-pointer",0,
-#endif 
+#endif
   "mips","-Os","-fomit-frame-pointer",0,
   "mipsel","-Os","-fomit-frame-pointer","-mno-abicalls","-fno-pic","-G","0",0,
   "ppc","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
